### DISCUSSION TITLE
Expanding compatibility beyond gnome ?

### DISCUSSION THREAD

**Discussion by le-patenteux** [gTile/gTile#395] [General]
Hello all.

First-off, thank you for this amazing piece of software. I have been using it for several years now as I own a 32:9 Ultrawide monitor and gTile is the only way I can manage windows effectively on my computer.

My issue is the following: Pop!OS, my distro of choice, and the one my entire team is using at work, is about to move out of gnome and switch to it's new Cosmic Desktop. gTile, being a Gnome extension, will not be compatible anymore!

Is there a way to get gTile to work on other DE than Gnome? Is there a fork of gTile for other DE ?

Regards.

**Comment by schnz** [comment]
Hi @Normand-Nadon,

the short answer is: To my knowledge, there is no fork of gTile for other DE's. A port, in general, is possible, albeit it would require a volunteer to do it.

Longer answer: As you might know, gTile was completely rewritten for Gnome 45 since some of the APIs changed and ES module import syntax is now required. When I did the rewrite, I explicitly had your use-case in mind and I tried to organize the code such, that a port to a different desktop environment should be possible with _relative_ ease. However, depending on how much the other DE diverges from Gnome it might still take a significant amount of work to make the port work.  

To be honest, its the first time I heard of the Cosmic DE, and I can't really find a reference to a documentation on how to write extensions for it (if at all possible) - maybe you have a reference? However, in case that there should be a serious effort behind porting gTile to any other DE, I would probably try to move those parts of gTile that are agnostic to the desktop environment  to a separate repository. It would require to define interfaces for all the dependencies (which are currently Gnome specific). But other than that, it shouldn't be too much of an effort. However, since I don't have any experience with the extension ecosystem of other DE's, I've no idea whether that approach makes sense or if a simple one-time fork is the superior approach. 

---

Related: #103 #20

**Comment by le-patenteux** [comment]
> To be honest, its the first time I heard of the Cosmic DE, and I can't really find a reference to a documentation on how to write extensions for it (if at all possible) - maybe you have a reference? However, in case that there should be a serious effort behind porting gTile to any other DE, I would probably try to move those parts of gTile that are agnostic to the desktop environment to a separate repository. It would require to define interfaces for all the dependencies (which are currently Gnome specific). But other than that, it shouldn't be too much of an effort. However, since I don't have any experience with the extension ecosystem of other DE's, I've no idea whether that approach makes sense or if a simple one-time fork is the superior approach.

Thanks for taking the time to answer me!
The DE is expected to release this spring... There is not a lot of documentation on it yet.
Pop!OS used to rely on gnome + some customisations added on top but they decided to stop using gnome starting from the next release for some reasons. Their new DE is programmed in Rust and aims to be a stable long-term target for them.

This is the public repo for it:
https://github.com/pop-os/cosmic-epoch

**Comment by scherepanov** [comment]
There is an identically named extension gTile for Cinnamon. It is written from scratch by independent developer, and has subset of functionality comparing to this project. No code has been re-used.

About support of other DE.

Yes Linus is open source, anybody can do any project at will.
Problem with new large scale projects like new DE is that it is fragmenting user and developer base and lead to lower quality of all DE.
I am not trying to say that Gnome is perfect, but it works and it is in decent shape. My personal opinion is to stick with Gnome and make gTile to work perfectly there.

gTile project never had a goal to be portable between different DE, it is/was very Gnome-centric. Redoing codebase to make it portable to various DE is a huge project, similar to complete rewrite that just been done. I am not sure it is a good spending of developer resources. Probably complete rewrite for each DE is better approach.

**Comment by schnz** [comment]
> This is the public repo for it:
https://github.com/pop-os/cosmic-epoch

That's what I also found during a quick lookup - but there doesn't seem to be any documentation regarding as to how extensions are supposed to be working (if at all). That would be the most important thing to have in order to make a judgement as to whether a port is possible _and_ reasonable.

> gTile project never had a goal to be portable between different DE, it is/was very Gnome-centric. Redoing codebase to make it portable to various DE is a huge project, similar to complete rewrite that just been done. I am not sure it is a good spending of developer resources. Probably complete rewrite for each DE is better approach.

While that might have been true for the pre-rewrite version of gTile, this is not necessarily true for the current code base:

- There are certain parts that are already DE agnostic (basically everything in `src/util`.
- Other parts (essentially `src/core`) are essentially re-usable across DE's after a little refactoring, i.e., working with interfaces as dependencies as opposed to directly using Gnome-specific instances/interfaces. 
- The UI part (`src/ui`) might require a complete and independent rewrite OR it might be re-usable with minimal effort  depending on whether the DE is based on the Gnome UI libraries (Clutter, St, Gtk, ...)

**Comment by le-patenteux** [comment]
Thank you so much for the time you spend explaining this to me. I am an advanced Linux user, but when it comes to the intricacies of how the DE in conjunction with the display manager and all the sorcery behind that makes a window appear where it does... I have no clue!

I will poke around the Cosmic repo and try to link this conversation there to spark interest

Have a nice day

**Comment by JMurph2015** [comment]
Hi!
I think I should probably chime in here, as I _have_ looked into this issue before.  With regards to Cinnamon, the extension there is based on an extremely old version of gTile which was manually ported to work with Cinnamon, I believe.  This works as well as it does because Cinnamon itself fundamentally descends from Gnome 2.x, and thus it has/had a pretty similar extension system.

For PopOS COSMIC DE (the Rust clean rewrite one), the odds that gTile will work there at all are pretty close to none.  GNOME is pretty unique situation in that GJS (the GNOME JavaScript runtime) is a core part of the desktop environment.  It introduces a lot of unpredictable behavior and resource usage that has caused problems over the years (see back to the great memory leak of circa 2019).  For those reasons, I doubt that COSMIC will reintroduce the same system for extensions.  If the extension language is not JavaScript/TypeScript then the gTile codebase won't work there no matter how cleanly we abstract the interfaces.

I think it *is* useful to refactor gTile so that we can be generic to "GNOME-like" extension systems like Cinnamon because it will let us offer a more up-to-date version for that DE and will make maintenance for future GNOME changes substantially easier.
Cheers!
